agent/consul/grpc_resolver.go

```diff
@@ -97,17 +181,38 @@ func (s *ServerResolverBuilder) getAddrs() []resolver.Address {
 // ServerResolver is a grpc Resolver that will keep a grpc.ClientConn up to date
 // on the list of server addresses to use.
 type ServerResolver struct {
-	cc            resolver.ClientConn
+	clientConn    resolver.ClientConn
 	closeCallback func()
+
+	lastAddrs []resolver.Address
+	addrLock  sync.Mutex
 }
 
 // updateAddrs updates this ServerResolver's ClientConn to use the given set of addrs.
 func (r *ServerResolver) updateAddrs(addrs []resolver.Address) {
-	r.cc.NewAddress(addrs)
+	// Only pass the first address initially, which will cause the
+	// balancer to spin down the connection for its previous first address
+	// if it is different. If we don't do this, it will keep using the old
+	// first address as long as it is still in the list, making it impossible to
+	// rebalance until that address is removed.
+	var firstAddr []resolver.Address
+	if len(addrs) > 0 {
+		firstAddr = []resolver.Address{addrs[0]}
+	}
+	r.clientConn.UpdateState(resolver.State{Addresses: firstAddr})
+
+	// Call UpdateState again with the entire list of addrs in case we need them
+	// for failover.
+	r.clientConn.UpdateState(resolver.State{Addresses: addrs})
```

Comment:
I wonder how reliable this is? It seems like we are relying on the implementation being synchronous such that it will have taken down the current conn before the UpdateState call returns. If that wasn't true might we still fail to force the reconnection in some cases?

I wonder how  likely it is to change behaviour in the future?

Would it be safer to implement our own balancer instead of using pick_first which is a copy paste? Or is this in a lower level than the balancer/picker that it keeps connections if the address is still OK?

Reply:
The subConn addr update call is actually synchronized with a lock so it **is** safe to call it in sequence that way, though it definitely feels jank to do it like this. 

> Or is this in a lower level than the balancer/picker that it keeps connections if the address is still OK?
That's right, the logic is in the addrConn and not the balancer, so we'd just be pushing this logic into a custom balancer which doesn't seem like it would change anything since addrConn isn't exposed publicly.

Reply:
Got it. I guess this is best we can do for now then!

Reply:
I wonder if it's possible to test the connection dropping behaviour in a non-brittle way so we could actually notice if the gRPC implementation changes in the future and breaks this assumption?

The grpc tests themselves seem to set up a bunch of test servers in-process and then run scenarios and assert about them. I guess we could do the same although they would need to be more property-based than rely on exact deterministic runs due to shuffle and timing etc.

We could for example though have it connect to one of three servers, figure out which one by inspecting the state of those test servers, then intentionally remove that one from the pool somehow and verify it picks up another? I guess we'd need to be able to control the shuffle and delivery of the server ips though a little more than just running full consul clients and letting serf do it?

We can consider that for later perhaps.

Reply:
The tests here do just that: https://github.com/hashicorp/consul/pull/6625/files#diff-a7649ae5f6cc2d79079d07fe312dd41fR73-R210
Unless you meant something different?